File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/python/04.M coloring problem.py
"""
=============================================================================
  QUESTION: M-Coloring Problem (GFG)
=============================================================================

  Can graph be colored with at most M colors? No two adjacent same color.

  APPROACH: Backtracking — try colors 1..M for each node.
  is_possible: check no neighbor has same color.

  DRY RUN triangle graph (0-1,1-2,0-2), M=3:
    Node 0→color 1, Node 1→color 1 CLASH→color 2,
    Node 2→color 1 CLASH→color 2 CLASH→color 3 → [1,2,3] ✓
  M=2: All combinations fail (triangle needs 3 colors) → False ✓

  Time: O(M^N), Space: O(N)
=============================================================================
"""


def is_possible(graph, color, n, col, node):
    for k in range(n):
        if k != node and graph[node][k] == 1 and color[k] == col:
            return False
    return True


def solve(graph, m, n, color, node):
    if node == n:
        return True

    for i in range(1, m + 1):
        if is_possible(graph, color, n, i, node):
            color[node] = i
            if solve(graph, m, n, color, node + 1):
                return True
            color[node] = 0  # Backtrack

    return False


def graph_coloring(graph, m, n):
    color = [0] * n
    return solve(graph, m, n, color, 0)


# Driver Code
if __name__ == "__main__":
    graph = [
        [0, 1, 1, 1],
        [1, 0, 1, 0],
        [1, 1, 0, 1],
        [1, 0, 1, 0]
    ]
    print(graph_coloring(graph, 3, 4))  # True
    print(graph_coloring(graph, 2, 4))  # False

File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/javascript/04.M coloring problem.js
/*
=============================================================================
  QUESTION: M-Coloring Problem (GFG)
=============================================================================

  Given undirected graph (adjacency matrix) and M colors, determine if
  graph can be colored with at most M colors such that no two adjacent
  vertices share the same color.

  Example: 4-node graph, M=3 → true | M=2 → false

=============================================================================
  APPROACH: Backtracking — Try Colors for Each Node — O(M^N) Time, O(N) Space
=============================================================================

  For each node: try colors 1 to M.
  - If safe (no neighbor has same color) → assign, recurse for next node.
  - If recursion fails → backtrack (reset to 0), try next color.
  - All nodes colored → return true. No color works → return false.

  Dry Run (triangle 0-1-2, M=3):
    node 0: color 1 ✓ → node 1: color 1 ✗ (adj to 0), color 2 ✓
      → node 2: color 1 ✗ (adj to 0), color 2 ✗ (adj to 1), color 3 ✓ → true ✓

=============================================================================
*/

function graphColoring(graph, m, n) {
    let color = new Array(n).fill(0);

    function isSafe(node, col) {
        for (let k = 0; k < n; k++)
            if (graph[node][k] === 1 && color[k] === col) return false;
        return true;
    }

    function solve(node) {
        if (node === n) return true;

        for (let c = 1; c <= m; c++) {
            if (isSafe(node, c)) {
                color[node] = c;
                if (solve(node + 1)) return true;
                color[node] = 0;  // Backtrack
            }
        }
        return false;
    }

    return solve(0);
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
let graph = [
    [0, 1, 1, 1],
    [1, 0, 1, 0],
    [1, 1, 0, 1],
    [1, 0, 1, 0]
];
console.log(graphColoring(graph, 3, 4));  // true
console.log(graphColoring(graph, 2, 4));  // false

